hazelcast/src/main/java/com/hazelcast/spi/impl/AbstractInvocationFuture.java

```diff
@@ -357,11 +357,12 @@ public final boolean complete(Object value) {
         for (; ; ) {
             final Object oldState = state;
             if (isDone(oldState)) {
-                // it can be that this invocation future already received an answer, e.g. when an invocation already received a
-                // response, but before it cleans up itself, it receives a HazelcastInstanceNotActiveException.
-                logger.warning("The Future.complete(Object value) can only be called once. Request: " + invocationToString()
-                        + ", current response: " + getState() + ", new response: " + value);
-
+                if (oldState != value) {
+                    // it can be that this future already completed, e.g. when an invocation already
+                    // received a response, but before it cleans up itself, it receives a HazelcastInstanceNotActiveException.
+                    logger.warning("Future.complete(Object value) can only be called once. Request: " + invocationToString()
+                            + ", current value: " + getState() + ", offered value: " + value);
```

Comment:
Volatile reads are cheap and compared to the price of logging, it will hardly make any impact. But it is nicer from a coding point of view since it causes less confusion.